lib/Demangling/Punycode.cpp

```diff
@@ -310,60 +314,74 @@ static bool convertUTF8toUTF32(llvm::StringRef InputUTF8,
       }
     } else if (first < 0xC0) {
       // Invalid continuation byte.
-      return false;
+      isInvalid = true;
     } else if (first < 0xE0) {
       // Two-byte sequence.
-      if (ptr == end)
-        return false;
-      uint8_t second = *ptr++;
-      if (!isContinuationByte(second))
-        return false;
-      code_point = ((first & 0x1F) << 6) | (second & 0x3F);
+      if (ptr == end) {
+        isInvalid = true;
+      } else {
+        uint8_t second = *ptr++;
+        if (!isContinuationByte(second))
+          isInvalid = true;
```

Comment:
I _think_ this is wrong. [The Unicode Standard](https://www.unicode.org/versions/Unicode16.0.0/core-spec/) says, in [3.9.5 Constraints on Conversion Processes](https://www.unicode.org/versions/Unicode16.0.0/core-spec/chapter-3/#G48534):

> If the converter encounters an ill-formed UTF-8 code unit sequence which starts with a valid first byte, but which does not continue with valid successor bytes (see [Table 3-7](https://www.unicode.org/versions/Unicode16.0.0/core-spec/chapter-3/#G27506)), it must not consume the successor bytes as part of the ill-formed subsequence whenever those successor bytes themselves constitute part of a well-formed UTF-8 code unit subsequence.

This means we _should not_ be consuming the second byte in that case, but we clearly are.  It goes on to give an example, namely that `C2 41 42` should be understood as `U+FFFD, U+0041, U+0042` rather than `U+FFFD` or `U+FFFD, U+0042`.

The other cases (three-byte and four-byte sequences) have the same problem.